### ISSUE TITLE
Validate AST and Specify API

### ISSUE THREAD

**Issue by jgoppert** [pymoca/pymoca#102]
I think it would be good to start the AST API discussion. I'm finding myself expanding array expressions for the xml backend which I know has already been done for the casadi backend. It would be nice if this was handeled on the AST side.

If we look at the ModelicaXML tree descriptions for a guide. I think it would make sense to be able to ask  the pymoca API for the tree at any stage 1, 2, 3 or 4. This would let us share the array expansion work between backends, which is my current motivation.

> 1. Modelica source code. This format is an abstract representation of Modelica code and should provide a one to one mapping between Modelica source code and a corresponding XML representation.
> 2. Instantiated hierarchical models. In this model representation, Modelica modifiers have taken effect, and the hierarchical structure is preserved.
> 3. Flattened Modelica models. This model representation is flat in the sense that it does not contain hierarchical model elements. It may, however, contain array declarations.
> 4. Flattened transformed models. In this representation, array variables and equations have been scalarized, alias variables have been eliminated and DAE index has been reduced, if needed.

Maybe something like this?
```python
# step 1
ast_code = parse(pymola)
generate_xml_code(ast_code).validate()

# step 2
ast_inst = instantiate(ast_code)
generate_xml_inst(ast_int).validate()

# step 3
ast_flat = flatten(ast_inst)
generate_xml_flat(ast_flat).validate()

# step 4
ast_reduced = reduce(ast_flat)
generate_xml_reduced(ast_reduced).validate()

# send to backends
generate_casadi(ast_flat... or ast_reduced, whatever is easiest)
```

**Comment by jbaayen** [comment]
For AST manipulation, I'd like to see equivalents for at least the following:

* `ast.Class.{add,remove}_class(cls: ast.Class)`
* `ast.Class.{add,remove}_symbol(sym: ast.Symbol)`
* `ast.Class.{add,remove}_equation(equation: ast.Equation)`

**Comment by jackvreeken** [comment]
These steps are currently implemented in two places: The flattener/`tree.py` handles (2) instantiation and (3) flattening, and the CasADi backend/`model.py` handles the array expansion. Do I understand you correctly that you would want to move step 4 to be _before_ handing it off to the backend? I think it is going to be tough to do alias detection/for loops/etc without relying on a symbolic framework like CasADi.

As far as step 2 and 3 are concerned, they are called/generated (not) exactly that way in https://github.com/pymoca/pymoca/blob/a762d3315691d84d49610df925c60c367c975303/src/pymoca/tree.py#L787 (constant thing is.. iffy).

**Comment by jgoppert** [comment]
Yes, that was my intent. If we had a function to expand array equations in tree.py it would avoid some code duplication in the backends if we could pull it off. I just paused before rewriting it for the xml backend to see of there was a better way.

**Comment by jackvreeken** [comment]
I very much agree it would be nice to have array expansion(/alias detection/DAE index reduction) shared like that. I have a feeling it will then have to be a listener/visitor like the backends are, but instead of AST->CasADi/XML it would do AST->AST. 

That said, I'm afraid of complications that would arise with:
- For/while loops
- Function definitions and calls
- If-else statements
- ...

Honestly, I can't quite predict what it will look like, what issues might pop up, and how it might influence the existing backends. Maybe we just have to try and see?

**Comment by jgoppert** [comment]
@jackvreeken @jbaayen I went down a rabbit hole trying to find a clean way to do make sure we have a consistent AST and validate it. I investigated writing code generation for the AST bindings, then looked into binding libraries. I wasn't happy with anything python offered, but found a very interesting BSD compatible C++ library (https://www.codesynthesis.com/products/xsd). This led me to try out the antlr C++ runtime with our grammar. It is kind of tempting.

## Cymoca
* https://github.com/jgoppert/cymoca
 
### Upside
* The AST is completely generated from the ModelicaXML.xsd
* The parser is completely genereated from Modelica.g4
* There is only one class we have to write, that does the mapping between the two, very little code maintenance. 
* This code looks very similar to our python backend. https://github.com/jgoppert/cymoca/blob/master/src/cymoca/Compiler.cpp#L53
* Faster compilation, lower memory footprint
* Bidirectional XML read/write, but stored as a C++ class.
* It is impossible to generate a non-valid ModelicaXML document, since the AST is a direct mapping.
* A python binding like Casadi uses would be possible.

### Downside
* Changing languages is a pretty huge deal, so I'm not sure how rapidly we could port existing code.
* I'm not sure how this works for your use-case
* C++ can be less hack-able, requires more setup time.

**Comment by jackvreeken** [comment]
As far as XSD->* is concerned, a quick google turned up PyXB. They say they are somewhat similar to CodeSynthesis (but then of course aimed at Python). Did you take a look at this one?

Changing the language would be a very big deal indeed, and not very high on my priority list. Going to C/C++ does open up other capabilities though (linking to _everything_), so it's not out of the question.

**Comment by jgoppert** [comment]
Yeah, I checked out PyXB. I keep finding myself trying to check the types of everything in python (I think I'm a C++ dev at heart). I have a hard time swallowing duck typing in the compiler. PyXB doesn't generate type names, so it is a pain and none of the auto-completion for PyCharm works. I did find monkeytype, which will go through and annotate python code with types and that sort of works as long as you get full coverage. But at that point I was asking myself if I'm spending so much time with type checking maybe C++ would be an alternative. 

I'll keep cymoca as a side project and see if it pans out for some of my use cases.

**Comment by jbaayen** [comment]
What about

http://json-schema.org/

https://pypi.python.org/pypi/jsonschema

and doing validation on our current AST in Python?

**Comment by jbaayen** [comment]
Note that we used the C++ XSD library you link to for RTC-Tools 1.x. It wasn't always easy to have it compile with all the C++ compilers we used (clang on Mac, gcc on Linux, and VS on Windows).

**Comment by jbaayen** [comment]
Result of today's weekly meeting: We stick with our in-memory AST for now, and work on finalizing the XML backend with XSD validation.

**Comment by jgoppert** [comment]
This is kind of crazy, but I wrote a g4 gramamr to XSD compiler. https://github.com/jgoppert/antlr2xsd It compiles [Modelica.g4](https://github.com/jgoppert/antlr2xsd/blob/master/test/g4/Modelica.g4) into [Pymoca.xsd](https://github.com/jgoppert/antlr2xsd/blob/master/test/output/Pymoca.xsd), example file here: [Test.xml](https://github.com/jgoppert/antlr2xsd/blob/master/test/xml/Test.xml). I've done the hard part of figuring out the multiplicity, but need to get the lexical tokens in there and some more stuff listed on the roadmap. Hopefully I can wrap it up this weekend to where it useful.

I figure that if we are serious about representing the source code with XML, the grammar is the best way to do it.

I also tried manual creation, but found it tedious and design deicision such as whether or not to use choices or polymorphism and the impact on code generation stopped me from wasting too much time on it.

I think the plan at this point is get the XSD for the source working well, and then manually define what the tree looks like after transformation using the automatically created elements.

**Comment by jgoppert** [comment]
It didn't end up working out well. It required too many special cases to clean stuff up. So I focused back on the manual XSD creation to match the grammar. That is wrapping up now.. I will replace the ModelicaXML generator with the new Pymoca.xsd  and then we can start validating the various stages in compilation. I think at that point, we will be able to print the xml easily and maybe printing source code at that point won't be much harder. Then if we find that the current AST needs some tweaking, we can do it then.

**Comment by jgoppert** [comment]
@jbaayen This is the current state, let me know if you have any feedback. https://github.com/jgoppert/pymoca_xsd/blob/master/Pymoca.xsd I tried to stick to ModelicaXML like syntax where possible, but follow the grammar more so we can attempt round trip from source to AST to source and pretty print.

**Comment by jbaayen** [comment]
Looks nice. Do we want componentrefs to be strings, or use a nesting construct there?

**Comment by jgoppert** [comment]
Some of that I glossed over on the first past. I will make sure it can handle subscripts etc. like in the grammar.

**Comment by jgoppert** [comment]
@jbaayen does this look like it will work?

```xml
<ref name="Some">
  <sub>0</sub>
  <slice/>
  <ref name="Stuff">
    <sub>1</sub>
    <sub>2</sub>
    <sub>3</sub>
  </ref>
</ref>
```

Representing Sum[0,:].Stuf[1,2,3]

**Comment by jbaayen** [comment]
Not sure about distinguishing `<subref>` and `<ref>`. Nesting refs sounds cleaner to me, with `<index>` children corresponding to each index in order. The `<index>` elements could then contain primitives, expressions, slices, and/or other component refs.

**Comment by jbaayen** [comment]
Regarding the component declarations: Nice idea to group by type. We lose quite a bit of verbosity like that.

**Comment by jgoppert** [comment]
On the subref/ ref thing, the issue is that I declare ref as an option for a primary, but I can name subref anything, because that indicates the subcomponent. I think I just need a better name for subref.

With the component declarations I'm just trying to match the grammar. So my thought was if I didn't handle that and did it the ModelicaXML on the round trip the source wouldn't be maintained.

**Comment by jgoppert** [comment]
Is this what you are suggesting regarding index? @jbaayen 
```xml
        <ref name="Some">
          <index>
            <add>
              <int>1</int>
              <int>1</int>
            </add>
            <ref name="x"/>
            <slice/>
          </index>
          <subref name="Stuff">
            <int>1</int>
            <slice/>
          </subref>
        </ref>
```

**Comment by jbaayen** [comment]
I was thinking along these lines:

```xml
        <ref name="Some">
          <index>
            <add>
              <int>1</int>
              <int>1</int>
            </add>
          </index>
          <index>
            <ref name="x"/>
          </index>
          <index>
            <slice/>
          </index>
          <ref name="Stuff">
            <index>
              <int>1</int>
            </index>
            <index>
              <slice/>
            </index>
          </ref>
        </ref>
```

**Comment by jbaayen** [comment]
The reason being that like this, it's clear to the reader that the various elements (add, ref, slice, ...) refer to indices.

It wasn't clear to me on first read, hence the idea :)
